### DISCUSSION TITLE
Infer type in namedtuple constructor

### DISCUSSION THREAD

**Discussion by GuillaumeDesforges** [microsoft/pyright#5754] [Ideas]
`namedtuple` is an interesting way to quickly build an immutable product type in Python.
However it seems that type inference in pyright does not apply when using them.

Take this simple example

```python
from collections import namedtuple
from typing import Protocol


class SomeProtocol(Protocol):
    def some_method(self, some_arg: int):
        ...


some_impl: SomeProtocol = namedtuple(
    "SomeImplementation",
    ["some_method"]
)(
    some_method=lambda: None,
)

some_impl.some_method(some_arg=1)
```

This program is not correct:

```
Traceback (most recent call last):
  File "/home/me/scratchpad.py", line 17, in <module>
    some_impl.some_method(some_arg=1)
TypeError: <lambda>() got an unexpected keyword argument 'some_arg'
```

The fix here is to do this

```diff
-     some_method=lambda: None,
+    some_method=lambda some_arg: None,
```

My point here is that, in the incorrect case, pyright does not display any error.

From my (uneducated) point of view, I would think that it should be possible to use external constraints (here the protocol type bound on `some_impl` value) to catch that the attribute `some_method` defined in the namedtuple constructor can't be bound to the type of `some_method` in `SomeProtocol`.

As alternatives, I considered using a frozen dataclasse, but I can't find a way to properly define attributes with a type of a complex callable (even using Protocol with a definition of __call__).

What do you think about this? Does that sound interesting? feasible?

**Comment by GuillaumeDesforges** [comment]
Note that, technically I can get type safety by expliciting types.

```
from typing import Any, NamedTuple, Protocol


class SomeProtocol(Protocol):
    @staticmethod
    def some_method(some_arg: int):
        ...


class SomeMethod(Protocol):
    def __call__(self, some_arg: int) -> Any:
        ...


class SomeImplementation(NamedTuple):
    some_method: SomeMethod


some_impl: SomeProtocol = SomeImplementation(
    some_method=lambda some_arg: None,
)

some_impl.some_method(some_arg=1)
```

However, having type inference on `namedtuple` would remove a lot of boilerplate and make everything more enjoyable.

**Reply by erictraut** [reply]
The way you're using `namedtuple` here is very unusual. You're creating a new class and immediately instantiating it in the same expression. I've never seen this done before, so I'm not inclined to add complex special-case logic to infer types in this case.

If you want to create typed named tuples, there's already a well-defined way to do this in Python, through the use of `typing.NamedTuple`. You can use it in two forms — in a `class` statement (which I prefer because it's cleaner and easier to read) and in a functional form.

```python
SomeImplementation = NamedTuple(
    "SomeImplementation", [("some_method", SomeProtocol)]
)
```

or
```python
class SomeImplementation(NamedTuple):
    some_method: SomeProtocol
```

Pyright and other Python type checkers fully support `NamedTuple`.

**Reply by GuillaumeDesforges** [reply]
> You're creating a new class and immediately instantiating it in the same expression.

Sorry if that was confusing. That was not really the point for me.

I guess another way to explain my point is that if I write

```python
def foo(x: int) -> str:
    ...

bar = (foo,)
# code that calls bar
```

Then `bar[0]` has an **inferred** type `(x: int) -> str`.

My question is, given the snippet

```
def foo(x: int) -> str:
    ...

SomeNamedTuple = namedtuple("SomeNamedTuple", "baz")
bar = SomeNamedTuple(
    baz=foo,
)
```

would it be possible to have the type of `bar.baz` inferred in the same way?

With `NamedTuple`, one has to write a lot of boilerplate code.

```python
def foo(x: int) -> str:
    ...


class Baz(Protocol):
    def __call__(self, x: int) -> str:
        ...


SomeNamedTuple = NamedTuple(
    "SomeNamedTuple",
    [("baz", Baz)],
)
bar = SomeNamedTuple(
    baz=foo,
)
```

But this requires to write a Protocol for each member of the named tuple.

**Reply by erictraut** [reply]
If I understand your question correctly, you want to define a named tuple class that accepts different types. This is the job for generics.

```python
from typing import Generic, NamedTuple, TypeVar

def foo(x: int) -> str:
    ...

T = TypeVar("T")

class SomeNamedTuple(NamedTuple, Generic[T]):
    baz: T

bar = SomeNamedTuple(baz=foo)
reveal_type(bar.baz) # (x: int) -> str
```

**Reply by GuillaumeDesforges** [reply]
This is not my question.

My question is whether it woud be possible to infer the type of namedtuple components, the same way tuple components can have their type inferred, maybe using bidirectional inference.

**Reply by GuillaumeDesforges** [reply]
The doc on type inference ([ref](https://github.com/microsoft/pyright/blob/bd548692fb4ed73d25710956c322d2f66d75c327/docs/type-inference.md)) says

> #### Tuple Expressions
>
>When inferring the type of a tuple expression (in the absence of bidirectional inference hints), Pyright assumes that the tuple has a fixed length, and each tuple element is typed as specifically as possible.
>
>```python
># The inferred type is tuple[Literal[1], Literal["a"], Literal[True]]
>var1 = (1, "a", True)
>
>def func1(a: int):
>    # The inferred type is tuple[int, int]
>    var2 = (a, a)
>
>    # If you want the type to be tuple[int, ...]
>    # (i.e. a homogeneous tuple of indeterminate length),
>    # use a type annotation.
>    var3: tuple[int, ...] = (a, a)
>```

I'm wondering if this could be applied to namedtuple

**Reply by erictraut** [reply]
No, type inference doesn't apply in this case. When you call `namedtuple`, you're defining a class. That class has a defined type, and its members have defined types (in this case, they have the type `Any` because you are using the old-style `namedtuple` call rather than the type-friendly `NamedTuple` call).

If you want to define a named tuple class that supports arbitrary generic types, you can use the method I outlined above,

**Reply by GuillaumeDesforges** [reply]
Got it, thanks for your answer.
